README.md
# Serverless Sns Sqs Lambda

[![serverless](http://public.serverless.com/badges/v3.svg)](http://www.serverless.com)
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat)](LICENSE)
![Github Actions Status](https://github.com/agiledigital/serverless-sns-sqs-lambda/workflows/Node.js%20CI/badge.svg?branch=master)
[![Type Coverage](https://img.shields.io/badge/dynamic/json.svg?label=type-coverage&prefix=%E2%89%A5&suffix=%&query=$.typeCoverage.atLeast&uri=https%3A%2F%2Fraw.githubusercontent.com%2Fagiledigital%2Fserverless-sns-sqs-lambda%2Fmaster%2Fpackage.json)](https://github.com/plantain-00/type-coverage)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/agiledigital/serverless-sns-sqs-lambda.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/agiledigital/serverless-sns-sqs-lambda/context:javascript)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
![npm](https://img.shields.io/npm/v/@agiledigital/serverless-sns-sqs-lambda)

This is a Serverless Framework plugin for AWS lambda Functions. Currently, it
is possible to subscribe directly to an SNS topic. However, if you want to
provide retry capability and error handling, you need to write a whole lot of
boilerplate to add a Queue and a Dead Letter Queue between the Lambda and the
SNS topic. This plugin allows you to define an sns subscriber with a `batchSize`
and a `maxRetryCount` as simply as subscribing directly to the sns topic.

# Table of Contents

- [Install](#install)
- [Setup](#setup)

## Install

Run `npm install` in your Serverless project.

`$ npm install --save-dev @agiledigital/serverless-sns-sqs-lambda`

Add the plugin to your serverless.yml file

```yml
plugins:
  - "@agiledigital/serverless-sns-sqs-lambda"
```

## Setup

Provide the lambda function with the snsSqs event, the plugin will add the AWS SNS topic and subscription, SQS queue and dead letter queue, and the role need for the lambda.

```yml
functions:
  processEvent:
    handler: handler.handler
    events:
      - snsSqs:
          name: TestEvent # Required - choose a name prefix for the event queue
          topicArn: !Ref Topic # Required - SNS topic to subscribe to
          omitPhysicalId: true # Optional - default value is false but recommended to be set to true for new deployments (see below)
          batchSize: 2 # Optional - default value is 10
          maximumBatchingWindowInSeconds: 10 # optional - default is 0 (no batch window)
          maxRetryCount: 2 # Optional - default value is 5
          kmsMasterKeyId: !GetAtt SQSQueueKey.Arn # optional - default is none (no encryption) - see Notes on Encryption section below
          kmsDataKeyReusePeriodSeconds: 600 # optional - AWS default is 300 seconds
          deadLetterMessageRetentionPeriodSeconds: 1209600 # optional - AWS default is 345600 secs (4 days)
          deadLetterQueueEnabled: true # optional - default is true
          lambdaSqsPermissionsEnabled: true # optional - default is true
          visibilityTimeout: 120 # optional (in seconds) - AWS default is 30 secs
          rawMessageDelivery: true # Optional - default value is true
          enabled: true # Optional - default value is true
          filterPolicy: # Optional - filter messages that are handled
            pets:
              - dog
              - cat

            # Overrides for generated CloudFormation templates
            # Mirrors the CloudFormation docs but uses camel case instead of title case
            #
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html
            mainQueueOverride:
              maximumMessageSize: 1024
              ...
            deadLetterQueueOverride:
              maximumMessageSize: 1024
              ...
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html
            eventSourceMappingOverride:
              sourceAccessConfigurations:
                - Type: SASL_SCRAM_256_AUTH
                  URI: arn:aws:secretsmanager:us-east-1:01234567890:secret:MyBrokerSecretName
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sns-subscription.html
            subscriptionOverride:
              region: ap-southeast-2

resources:
  Resources:
    Topic:
      Type: AWS::SNS::Topic
      Properties:
        TopicName: TestTopic

plugins:
  - "@agiledigital/serverless-sns-sqs-lambda"
```

### What is the omitPhysicalId option?

AWS allows you to omit the physical ID (queue name) for SQS resources.

In this case, the name is automatically generated by AWS based on the logical ID (the CloudFormation resource name).

This provides some benefits such as the ability to perform updates that require replacement, and not having to worry about the 80 character maximum length for queue names.

However, if you need to refer to the queue by name (which should be rare), rather than a CloudFormation reference, you might have to switch this off so that the name is stable.

This would be set to 'true' by default if this was the first version of the plugin. However, since the plugin is already in use and it could break existing stacks and may not be backwards compatible we have set this to 'false' by default. Switching this to 'true' on an existing stack has been tested and it seems to work OK, although it does create new queues and deletes the old ones. Switch the option on an existing stack at your own risk.

See also: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html

### Notes on Encryption

If you choose to encrypt your SQS queue, the SNS topic will not be able to send it any messages if you use a managed key (alias/aws/sqs). This is due to an AWS limitation.

See: https://aws.amazon.com/premiumsupport/knowledge-center/sns-topic-sqs-queue-sse-kms-key-policy/

You will need to create a CMK like the following:

```yaml
# To allow SNS to push messages to an encrypted queue, a CMK must be used
SQSQueueCMK:
  Type: AWS::KMS::Key
  Properties:
    KeyPolicy:
      Version: "2012-10-17"
      Id: key-default-1
      Statement:
        - Sid: Enable IAM User Permissions
          Effect: Allow
          Principal:
            AWS: !Join
              - ""
              - - "arn:aws:iam::"
                - !Ref "AWS::AccountId"
                - ":root"
          Action: "kms:*"
          Resource: "*"
        - Sid: Allow SNS publish to SQS
          Effect: Allow
          Principal:
            Service: sns.amazonaws.com
          Action:
            - kms:GenerateDataKey
            - kms:Decrypt
          Resource: "*"
```

and then reference it in the `snsSqs` config with the `kmsMasterKeyId` attribute.

```yaml
functions:
  processEvent:
    handler: handler.handler
    events:
      - snsSqs:
          # ...
          kmsMasterKeyId: !GetAtt SQSQueueKey.Arn
          # ...
```

`kmsMasterKeyId` can either be a key ID (simple string) or an ARN or reference to to ARN. Using !Ref on a key will return a key ID and is invalid, so you'll need to use GetAtt and reference the Arn property.

### CloudFormation Overrides

If you would like to override a part of the CloudFormation template
that is generated by this plugin, you can pass raw CloudFormation
to the override config options outlined above.

The configuration must be provided with camel case keys,
but apart from that, you can use the CloudFormation config
as specified by AWS.

For example, if you wanted to override the maximumMessageSize for the main queue
you could find the "MaximumMessageSize" config option in the [AWS documentation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html)
make the key camel case ("maximumMessageSize") and pass it into the override section:

```yaml
    events:
      - snsSqs:
          name: Example
          ...
          mainQueueOverride:
            maximumMessageSize: 1024
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See
[Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [2.1.0](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v2.0.1...v2.1.0) (2023-05-24)


### Features

* empty commit to force release of [#711](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/711) ([b090fa7](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/b090fa731aa9add47154634c3041df9ff1f96179))

## [2.0.1](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v2.0.0...v2.0.1) (2022-05-24)


### Bug Fixes

* add 'prefix' to config schema (fixes [#562](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/562)) ([4602a63](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/4602a63e5b6e726f8931f3c3a3239117d25fe266))

# [2.0.0](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v1.0.1...v2.0.0) (2022-05-13)


### Bug Fixes

* fix case sensitivity issue ([42e9675](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/42e967520705b0d36fbf8b7b8030d40985e34a3b))
* improve the handling of encrypted SQS queues (fixes [#555](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/555)) ([789ea78](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/789ea786e599aefd8df8d51f4cf8ca70f74810a4))


### Features

* upgrade to serverless v3 ([#540](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/540)) ([cf842f0](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/cf842f0b461594ea03df0dfee3ff7909d6e2c04b))


### BREAKING CHANGES

* If you have implemented workarounds to allow the lambda to subscribe to an encrypted SQS queue, you may get conflicts as the policy to allow the decryption is now added automatically

- If you provide an key ID, key ARN or reference to a key ARN to the `kmsMasterKeyId` attribute, the relevant 'kms:Decrypt' policy statement should be added automatically to allow the subscription to work correctly
* serverless v2 is no longer supported. It might still work, but bug fixes/new features will mostly be developed for and tested with serverless v3 (important security/bug fixes _may_ be back ported to v2 versions depending on the uptake of v3)

serverless v3 is now a peer dependency, you will get warnings if you are on earlier versions

## [1.0.1](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v1.0.0...v1.0.1) (2022-05-13)


### Bug Fixes

* allow physical ID to be omitted ([#444](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/444)) ([b22944f](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/b22944f5094a544fe7b4f1b0358d21967668b6e2))
* ensure name uniqueness ([#444](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/444)) ([83617d6](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/83617d6b98164cf14b14944a3a94e9dec906267e))

# [1.0.0](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v0.9.1...v1.0.0) (2022-01-07)


### Bug Fixes

* add another source of stage ([e9c1cb1](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/e9c1cb185706dca8ffcaf8b5b1a82b80512215c1)), closes [#432](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/432)


### Features

* ensure that work done for issue [#432](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/432) is published as a new major version ([6433354](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/643335489b6488d8beb4ed2755c60623b16cabe1))


### BREAKING CHANGES

* the way that stage is determined has changed which might cause the generated CF template to change, causing issues on the next deploy

## [0.9.1](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v0.9.0...v0.9.1) (2021-10-23)


### Bug Fixes

* remove unneccessary 'DependsOn' block ([9276523](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/9276523d01cb510b6fd2f685f121b1c42a86a77e))
* support custom role ARNs ([563c202](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/563c20267151efa7122e8a3f7a5df0ea49a25c37)), closes [#350](https://github.com/agiledigital/serverless-sns-sqs-lambda/issues/350)

# [0.9.0](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v0.8.0...v0.9.0) (2021-10-13)


### Features

* add fifo queue support ([b1122b0](https://github.com/agiledigital/serverless-sns-sqs-lambda/commit/b1122b0bc01da6c0e84a3f83892017e37c9d88f3))

# [0.8.0](https://github.com/agiledigital/serverless-sns-sqs-lambda/compare/v0.7.0...v0.8.0) (2021-06-29)



lib/serverless-sns-sqs-lambda.ts
import { JsonObject } from "type-fest";

// Future work: Properly type the file
/* eslint-disable @typescript-eslint/explicit-module-boundary-types */
/* eslint-disable @typescript-eslint/no-explicit-any */

/**
 * A regular expression that matches AWS KMS arns
 */
const kmsArnRegex = /^arn:aws:kms:.*:.*:key\/.+$/;

/**
 * Defines the structure of the config object
 * that is passed to the main functions to
 * generate the Serverless template.
 */
type Config = {
  name: string;
  topicArn: string;
  funcName: string;
  prefix: string;
  batchSize: number;
  maximumBatchingWindowInSeconds: number;
  maxRetryCount: number;
  kmsMasterKeyId: string;
  kmsDataKeyReusePeriodSeconds: number;
  deadLetterMessageRetentionPeriodSeconds: number;
  deadLetterQueueEnabled: boolean;
  enabled: boolean;
  fifo: boolean;
  visibilityTimeout: number;
  rawMessageDelivery: boolean;
  filterPolicy: any;
  readonly omitPhysicalId: boolean;

  mainQueueOverride: JsonObject;
  deadLetterQueueOverride: JsonObject;
  eventSourceMappingOverride: JsonObject;
  subscriptionOverride: JsonObject;
};

/**
 * Parse a value into a number or set it to a default value.
 *
 * @param {string|number|null|undefined} intString value possibly in string
 * @param {*} defaultInt the default value if `intString` can't be parsed
 */
const parseIntOr = (intString, defaultInt) => {
  if (intString === null || intString === undefined) {
    return defaultInt;
  }
  try {
    return parseInt(intString.toString(), 10);
  } catch {
    return defaultInt;
  }
};

/**
 * Converts a string from camelCase to PascalCase. Basically, it just
 * capitalises the first letter.
 *
 * @param {string} camelCase camelCase string
 */
const pascalCase = (camelCase: string): string =>
  camelCase.slice(0, 1).toUpperCase() + camelCase.slice(1);

const pascalCaseAllKeys = (jsonObject: JsonObject): JsonObject =>
  Object.keys(jsonObject).reduce(
    (acc, key) => ({
      ...acc,
      [pascalCase(key)]: jsonObject[key]
    }),
    {}
  );

const validateQueueName = (queueName: string): string => {
  if (queueName.length > 80) {
    throw new Error(
      `Generated queue name [${queueName}] is longer than 80 characters long and may be truncated by AWS, causing naming collisions. Try a shorter prefix or name, or try the hashQueueName config option.`
    );
  }
  return queueName;
};

/**
 * Returns true if the provided string looks like an KMS ARN, otherwise false
 * @param possibleArn the candidate string
 * @returns true if the provided string looks like a KMS ARN, otherwise false
 */
const isKmsArn = (possibleArn: string): boolean =>
  kmsArnRegex.test(possibleArn);

/**
 * Adds a resource block to a template, ensuring uniqueness.
 * @param template the serverless template
 * @param logicalId the logical ID (resource key) for the resource
 * @param resourceDefinition the definition of the resource
 */
const addResource = (
  template: any,
  logicalId: string,
  resourceDefinition: Record<string, unknown>
) => {
  if (logicalId in template.Resources) {
    throw new Error(
      `Generated logical ID [${logicalId}] already exists in resources definition. Ensure that the snsSqs event definition has a unique name property.`
    );
  }
  template.Resources[logicalId] = resourceDefinition;
};

/**
 * The ServerlessSnsSqsLambda plugin looks for functions that contain an
 * `snsSqs` event and adds the necessary resources for the Lambda to subscribe
 * to the SNS topics with error handling and retry functionality built in.
 *
 * An example configuration might look like:
 *
 *     functions:
 *       processEvent:
 *         handler: handler.handler
 *         events:
 *           - snsSqs:
 *             name: ResourcePrefix
 *             topicArn: ${self:custom.topicArn}
 *             batchSize: 2
 *             maximumBatchingWindowInSeconds: 30
 *             maxRetryCount: 2
 *             kmsMasterKeyId: alias/aws/sqs
 *             kmsDataKeyReusePeriodSeconds: 600
 *             deadLetterMessageRetentionPeriodSeconds: 1209600
 *             deadLetterQueueEnabled: true
 *             visibilityTimeout: 120
 *             rawMessageDelivery: true
 *             enabled: false
 *             fifo: false
 *             filterPolicy:
 *               pet:
 *                 - dog
 *                 - cat
 */
export default class ServerlessSnsSqsLambda {
  serverless: any;
  options: any;
  provider: any;
  custom: any;
  serviceName: string;
  stage: string;
  hooks: any;

  /**
   * @param {*} serverless
   * @param {*} options
   */
  constructor(serverless, options) {
    this.serverless = serverless;
    this.options = options;
    this.provider = serverless ? serverless.getProvider("aws") : null;
    this.custom = serverless.service ? serverless.service.custom : null;
    this.serviceName = serverless.service.service;

    // Aligns with AWS provider order of precedence: https://github.com/serverless/serverless/blob/46d090a302b9f7f4a3cf479695489b7ffc46b75b/lib/plugins/aws/provider.js#L1728
    // Serverless will set one of these to "dev" if it is not provided so we don't need an explicit fallback
    this.stage =
      this.options.stage ||
      this.serverless.config.stage ||
      this.serverless.service.provider.stage;

    serverless.configSchemaHandler.defineFunctionEvent("aws", "snsSqs", {
      type: "object",
      properties: {
        name: { type: "string" },
        topicArn: { $ref: "#/definitions/awsArn" },
        prefix: { type: "string" },
        omitPhysicalId: { type: "boolean" },
        batchSize: { type: "number", minimum: 1, maximum: 10000 },
        maximumBatchingWindowInSeconds: {
          type: "number",
          minimum: 0,
          maximum: 300
        },
        maxRetryCount: { type: "number" },
        kmsMasterKeyId: {
          anyOf: [{ type: "string" }, { $ref: "#/definitions/awsArn" }]
        },
        kmsDataKeyReusePeriodSeconds: {
          type: "number",
          minimum: 60,
          maximum: 86400
        },
        visibilityTimeout: {
          type: "number",
          minimum: 0,
          maximum: 43200
        },
        deadLetterMessageRetentionPeriodSeconds: {
          type: "number",
          minimum: 60,
          maximum: 1209600
        },
        deadLetterQueueEnabled: { type: "boolean" },
        rawMessageDelivery: { type: "boolean" },
        enabled: { type: "boolean" },
        fifo: { type: "boolean" },
        filterPolicy: { type: "object" },
        mainQueueOverride: { type: "object" },
        deadLetterQueueOverride: { type: "object" },
        eventSourceMappingOverride: { type: "object" },
        subscriptionOverride: { type: "object" }
      },
      required: ["name", "topicArn"],
      additionalProperties: false
    });

    if (!this.provider) {
      throw new Error("This plugin must be used with AWS");
    }

    this.hooks = {
      "aws:package:finalize:mergeCustomProviderResources":
        this.modifyTemplate.bind(this)
    };
  }

  /**
   * Mutate the CloudFormation template, adding the necessary resources for
   * the Lambda to subscribe to the SNS topics with error handling and retry
   * functionality built in.
   */
  modifyTemplate() {
    const functions = this.serverless.service.functions;
    const template =
      this.serverless.service.provider.compiledCloudFormationTemplate;

    Object.keys(functions).forEach(funcKey => {
      const func = functions[funcKey];
      if (func.events) {
        func.events.forEach(event => {
          if (event.snsSqs) {
            if (this.options.verbose) {
              console.info(
                `Adding snsSqs event handler [${JSON.stringify(event.snsSqs)}]`
              );
            }
            this.addSnsSqsResources(
              template,
              funcKey,
              this.stage,
              event.snsSqs
            );
          }
        });
      }
    });
  }

  /**
   *
   * @param {object} template the template which gets mutated
   * @param {string} funcName the name of the function from serverless config
   * @param {string} stage the stage name from the serverless config
   * @param {object} snsSqsConfig the configuration values from the snsSqs
   *  event portion of the serverless function config
   */
  addSnsSqsResources(template, funcName, stage, snsSqsConfig) {
    const config = this.validateConfig(funcName, stage, snsSqsConfig);

    [
      this.addEventSourceMapping,
      this.addEventDeadLetterQueue,
      this.addEventQueue,
      this.addEventQueuePolicy,
      this.addTopicSubscription,
      this.addLambdaSqsPermissions
    ].reduce((template, func) => {
      func(template, config);
      return template;
    }, template);
  }

  /**
   * Validate the configuration values from the serverless config file,
   * returning a config object that can be passed to the resource setup
   * functions.
   *
   * @param {string} funcName the name of the function from serverless config
   * @param {string} stage the stage name from the serverless config
   * @param {object} config the configuration values from the snsSqs event
   *  portion of the serverless function config
   */
  validateConfig(funcName, stage, config): Config {
    if (!config.topicArn || !config.name) {
      throw new Error(`Error:
When creating an snsSqs handler, you must define the name and topicArn.
In function [${funcName}]:
- name was [${config.name}]
- topicArn was [${config.topicArn}].

Usage
-----

  functions:
    processEvent:
      handler: handler.handler
      events:
        - snsSqs:
            name: Event                                      # required
            topicArn: !Ref TopicArn                          # required
            prefix: some-prefix                              # optional - default is \`\${this.serviceName}-\${stage}-\${funcNamePascalCase}\`
            maxRetryCount: 2                                 # optional - default is 5
            batchSize: 1                                     # optional - default is 10
            batchWindow: 10                                  # optional - default is 0 (no batch window)
            kmsMasterKeyId: alias/aws/sqs                    # optional - default is none (no encryption)
            kmsDataKeyReusePeriodSeconds: 600                # optional - AWS default is 300 seconds
            deadLetterMessageRetentionPeriodSeconds: 1209600 # optional - AWS default is 345600 secs (4 days)
            deadLetterQueueEnabled: true                     # optional - default is enabled
            enabled: true                                    # optional - AWS default is true
            fifo: false                                      # optional - AWS default is false
            visibilityTimeout: 30                            # optional - AWS default is 30 seconds
            rawMessageDelivery: false                        # optional - default is false
            filterPolicy:
              pet:
                - dog
                - cat

            # Overrides for generated CloudFormation templates
            # Mirrors the CloudFormation docs but uses camel case instead of title case
            #
            #
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-sqs-queues.html
            mainQueueOverride:
              maximumMessageSize: 1024
              ...
            deadLetterQueueOverride:
              maximumMessageSize: 1024
              ...
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-lambda-eventsourcemapping.html
            eventSourceMappingOverride:
              bisectBatchOnFunctionError: true
            # https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-sns-subscription.html
            subscriptionOverride:
              rawMessageDelivery: true

`);
    }

    const funcNamePascalCase = pascalCase(funcName);
    return {
      ...config,
      name: config.name,
      funcName: funcNamePascalCase,
      prefix:
        config.prefix || `${this.serviceName}-${stage}-${funcNamePascalCase}`,
      batchSize: parseIntOr(config.batchSize, 10),
      maxRetryCount: parseIntOr(config.maxRetryCount, 5),
      kmsMasterKeyId: config.kmsMasterKeyId,
      kmsDataKeyReusePeriodSeconds: config.kmsDataKeyReusePeriodSeconds,
      deadLetterMessageRetentionPeriodSeconds:
        config.deadLetterMessageRetentionPeriodSeconds,
      deadLetterQueueEnabled:
        config.deadLetterQueueEnabled !== undefined
          ? config.deadLetterQueueEnabled
          : true,
      enabled: config.enabled,
      fifo: config.fifo !== undefined ? config.fifo : false,
      visibilityTimeout: config.visibilityTimeout,
      rawMessageDelivery:
        config.rawMessageDelivery !== undefined
          ? config.rawMessageDelivery
          : false,
      mainQueueOverride: config.mainQueueOverride ?? {},
      deadLetterQueueOverride: config.deadLetterQueueOverride ?? {},
      eventSourceMappingOverride: config.eventSourceMappingOverride ?? {},
      subscriptionOverride: config.subscriptionOverride ?? {}
    };
  }

  /**
   * Add the Event Source Mapping which sets up the message handler to pull
   * events of the Event Queue and handle them.
   *
   * @param {object} template the template which gets mutated
   * @param {{funcName, name, prefix, batchSize, enabled}} config including name of the queue
   *  and the resource prefix
   */
  addEventSourceMapping(
    template,
    {
      funcName,
      name,
      batchSize,
      maximumBatchingWindowInSeconds,
      enabled,
      eventSourceMappingOverride
    }: Config
  ) {
    const enabledWithDefault = enabled !== undefined ? enabled : true;
    addResource(template, `${funcName}EventSourceMappingSQS${name}Queue`, {
      Type: "AWS::Lambda::EventSourceMapping",
      Properties: {
        BatchSize: batchSize,
        MaximumBatchingWindowInSeconds:
          maximumBatchingWindowInSeconds !== undefined
            ? maximumBatchingWindowInSeconds
            : 0,
        EventSourceArn: { "Fn::GetAtt": [`${name}Queue`, "Arn"] },
        FunctionName: { "Fn::GetAtt": [`${funcName}LambdaFunction`, "Arn"] },
        Enabled: enabledWithDefault ? "True" : "False",
        ...pascalCaseAllKeys(eventSourceMappingOverride)
      }
    });
  }

  /**
   * Add the Dead Letter Queue which will collect failed messages for later
   * inspection and handling.
   *
   * @param {object} template the template which gets mutated
   * @param {{name, prefix, kmsMasterKeyId, kmsDataKeyReusePeriodSeconds, deadLetterMessageRetentionPeriodSeconds }} config including name of the queue
   *  and the resource prefix
   */
  addEventDeadLetterQueue(
    template,
    {
      name,
      prefix,
      fifo,
      kmsMasterKeyId,
      kmsDataKeyReusePeriodSeconds,
      deadLetterMessageRetentionPeriodSeconds,
      deadLetterQueueOverride,
      deadLetterQueueEnabled,
      omitPhysicalId
    }
  ) {
    if (!deadLetterQueueEnabled) {
      return;
    }
    const candidateQueueName = `${prefix}${name}DeadLetterQueue${
      fifo ? ".fifo" : ""
    }`;
    addResource(template, `${name}DeadLetterQueue`, {
      Type: "AWS::SQS::Queue",
      Properties: {
        ...(omitPhysicalId
          ? {}
          : { QueueName: validateQueueName(candidateQueueName) }),
        ...(fifo ? { FifoQueue: true } : {}),
        ...(kmsMasterKeyId !== undefined
          ? {
              KmsMasterKeyId: kmsMasterKeyId
            }
          : {}),
        ...(kmsDataKeyReusePeriodSeconds !== undefined
          ? {
              KmsDataKeyReusePeriodSeconds: kmsDataKeyReusePeriodSeconds
            }
          : {}),
        ...(deadLetterMessageRetentionPeriodSeconds !== undefined
          ? {
              MessageRetentionPeriod: deadLetterMessageRetentionPeriodSeconds
            }
          : {}),
        ...pascalCaseAllKeys(deadLetterQueueOverride)
      }
    });
  }

  /**
   * Add the event queue that will subscribe to the topic and collect the events
   * from SNS as they arrive, holding them for processing.
   *
   * @param {object} template the template which gets mutated
   * @param {{name, prefix, maxRetryCount, kmsMasterKeyId, kmsDataKeyReusePeriodSeconds, visibilityTimeout}} config including name of the queue,
   *  the resource prefix and the max retry count for message handler failures.
   */
  addEventQueue(
    template,
    {
      name,
      prefix,
      fifo,
      maxRetryCount,
      kmsMasterKeyId,
      kmsDataKeyReusePeriodSeconds,
      visibilityTimeout,
      mainQueueOverride,
      omitPhysicalId,
      deadLetterQueueEnabled
    }: Config
  ) {
    const candidateQueueName = `${prefix}${name}Queue${fifo ? ".fifo" : ""}`;
    addResource(template, `${name}Queue`, {
      Type: "AWS::SQS::Queue",
      Properties: {
        ...(omitPhysicalId
          ? {}
          : { QueueName: validateQueueName(candidateQueueName) }),
        ...(fifo ? { FifoQueue: true } : {}),
        ...(deadLetterQueueEnabled
          ? {
              RedrivePolicy: {
                deadLetterTargetArn: {
                  "Fn::GetAtt": [`${name}DeadLetterQueue`, "Arn"]
                },
                maxReceiveCount: maxRetryCount
              }
            }
          : {}),
        ...(kmsMasterKeyId !== undefined
          ? {
              KmsMasterKeyId: kmsMasterKeyId
            }
          : {}),
        ...(kmsDataKeyReusePeriodSeconds !== undefined
          ? {
              KmsDataKeyReusePeriodSeconds: kmsDataKeyReusePeriodSeconds
            }
          : {}),
        ...(visibilityTimeout !== undefined
          ? {
              VisibilityTimeout: visibilityTimeout
            }
          : {}),
        ...pascalCaseAllKeys(mainQueueOverride)
      }
    });
  }

  /**
   * Add a policy allowing the queue to subscribe to the SNS topic.
   *
   * @param {object} template the template which gets mutated
   * @param {{name, prefix, topicArn}} config including name of the queue, the
   *  resource prefix and the arn of the topic
   */
  addEventQueuePolicy(template, { name, prefix, topicArn }: Config) {
    addResource(template, `${name}QueuePolicy`, {
      Type: "AWS::SQS::QueuePolicy",
      Properties: {
        PolicyDocument: {
          Version: "2012-10-17",
          Id: `${prefix}${name}Queue`,
          Statement: [
            {
              Sid: `${prefix}${name}Sid`,
              Effect: "Allow",
              Principal: { Service: "sns.amazonaws.com" },
              Action: "SQS:SendMessage",
              Resource: { "Fn::GetAtt": [`${name}Queue`, "Arn"] },
              Condition: { ArnEquals: { "aws:SourceArn": [topicArn] } }
            }
          ]
        },
        Queues: [{ Ref: `${name}Queue` }]
      }
    });
  }

  /**
   * Subscribe the newly created queue to the desired topic.
   *
   * @param {object} template the template which gets mutated
   * @param {{name, topicArn, filterPolicy}} config including name of the queue,
   *  the arn of the topic and the filter policy for the subscription
   */
  addTopicSubscription(
    template,
    {
      name,
      topicArn,
      filterPolicy,
      rawMessageDelivery,
      subscriptionOverride
    }: Config
  ) {
    addResource(template, `Subscribe${name}Topic`, {
      Type: "AWS::SNS::Subscription",
      Properties: {
        Endpoint: { "Fn::GetAtt": [`${name}Queue`, "Arn"] },
        Protocol: "sqs",
        TopicArn: topicArn,
        ...(filterPolicy ? { FilterPolicy: filterPolicy } : {}),
        ...(rawMessageDelivery !== undefined
          ? {
              RawMessageDelivery: rawMessageDelivery
            }
          : {}),
        ...pascalCaseAllKeys(subscriptionOverride)
      }
    });
  }

  /**
   * Add permissions so that the SQS handler can access the queue.
   *
   * @param {object} template the template which gets mutated
   * @param {{name, prefix}} config the name of the queue the lambda is subscribed to
   */
  addLambdaSqsPermissions(
    template,
    { name, kmsMasterKeyId, deadLetterQueueEnabled }
  ) {
    if (template.Resources.IamRoleLambdaExecution === undefined) {
      // The user has set their own custom role ARN so the Serverless generated role is not generated
      // We can safely skip this step because the owner of the custom role ARN is responsible for setting
      // this the relevant policy to allow the lambda to access the queue.
      return;
    }
    const queues = [{ "Fn::GetAtt": [`${name}Queue`, "Arn"] }];
    if (deadLetterQueueEnabled) {
      queues.push({ "Fn::GetAtt": [`${name}DeadLetterQueue`, "Arn"] });
    }
    template.Resources.IamRoleLambdaExecution.Properties.Policies[0].PolicyDocument.Statement.push(
      {
        Effect: "Allow",
        Action: [
          "sqs:ReceiveMessage",
          "sqs:DeleteMessage",
          "sqs:GetQueueAttributes"
        ],
        Resource: queues
      }
    );

    if (kmsMasterKeyId !== undefined && kmsMasterKeyId !== null) {
      // TODO: Should we rename kmsMasterKeyId to make it clearer that it can accept an ARN?
      const resource =
        // If the key ID is an object, it is most likely a "Ref" or "GetAtt" so we should pass it straight through so it gets resolved by CloudFormation
        // If an ARN is provided, pass it straight through too, because no processing is needed
        // Otherwise if it isn't either of those things, it is probably an ID, so we need to
        // transform it to an ARN to make the policy valid
        typeof kmsMasterKeyId === "object" || isKmsArn(kmsMasterKeyId)
          ? kmsMasterKeyId
          : `arn:aws:kms:::key/${kmsMasterKeyId}`;
      template.Resources.IamRoleLambdaExecution.Properties.Policies[0].PolicyDocument.Statement.push(
        {
          Effect: "Allow",
          Action: ["kms:Decrypt"],
          Resource: resource
        }
      );
    }
  }
}
